### DISCUSSION TITLE
Access to local files by Dask Gateway workers

### DISCUSSION THREAD

**Discussion by XingongLi** [microsoft/PlanetaryComputer#31] [Q&A]
I’m trying to use PC and Dask Gateway to speed up a flood mapping application. My data (uploaded to PC and saved under my account) are divided into tiles (in MATLAB .mat format) and each tile need to be mapped as a GeoTiff file and all the tile maps will be mosaiced when individual tile mapping is done. I can run the application as a notebook on PC without any problem. When I tried to parallelize the application using the Dask Gateway it gave me the error "FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/fldpln/libraries/spring/FLDPLN_tiled_10.mat'".  Below is the custom DAG for the parallelization. It seems to me that gateway workers cannot access those tile files and I suspect they also cannot write the maps. So my question is how I can make my local file system available to gateway workers?

![image](https://user-images.githubusercontent.com/17601071/152654643-65947b5f-53cb-43ab-af83-13ba486c7b27.png)

**Comment by TomAugspurger** [comment]
I'd recommend reading a bit about how Dask Gateway works: https://gateway.dask.org/. The machines running your computation are physically separate from the ones where you're client is running. They don't share a filesystem.

If you use something like an Azure Blob Storage container to have a location that all nodes can read or write to.

Or, to answer the specific question of 'How can I make *a* file available to the workers, there's https://distributed.dask.org/en/stable/api.html?highlight=local%20file#distributed.Client.upload_file.

**Comment by XingongLi** [comment]
Based on my communication with the dask discourse group, I should NOT use clinet.upload_file or client.scatter to distribute the files to workers. The gateway should have a way (or policy) to make client data available to the cluster. So my question is, for the gateway provided by the PC, how the client  can make their data accessible to the cluster? Is an Azure Blob Storage the only option? Any documents or examples on how to make this happen? Anyone from the team can help?

**Reply by TomAugspurger** [reply]
https://planetarycomputer.microsoft.com/docs/quickstarts/storage#Write-to-Azure-Blob-Storage

**Comment by XingongLi** [comment]
Can a PC user create a free Blob Storage account? And what's the storage limit if we can? 

My code uses scipy.sio.loadmat() to read local mat files. Assuming that I can upload the mat files to a storage account, how can the gateway workers read the mat files from the blob storage? I don't think loadmat() supports direct reading from blob storage.

**Reply by TomAugspurger** [reply]
You can download them locally to the worker's disk first if needed.

**Comment by XingongLi** [comment]
I have an organization Azure portal account but I cannot create a blob storage account with it. Azure won't let me add a new Resource group and set Region to (Europe) West Europe which is where PC located I believe. Should I create a personal Azure account and then create the storage account? Any free research storage account? Right now, I just want to test and see how the dask gateway can speed up flood mapping.

**Reply by TomAugspurger** [reply]
Either getting your organization to create you a storage account / container or using a personal account (with the free trial) sounds reasonable.

**Comment by XingongLi** [comment]
Tom,

I've got those things figured out. I changed my code and uploaded one small watershed to azure blob storage for testing. The parallelized flood mapping was running last night and used only half of the time of a linear mapping! I'll try a large watershed (I hope it won't reach my storage limit if there is one) and see the benefit of parallelization. I'm very new to Dask and would like to explore more on better the parallelization.

In addition to the flood mapping project, I also plan to do a global snow cover change analysis on PC using MODIS MOD10A1 & MYD10A1 data. I have done some before on Earth Engine (see TrendySnow<https://trendysnow.herokuapp.com/> web application built using EE as the backend engine). But there are limitations on using some trend analysis methods on EE. I think I may be able to do those analyses on PC with its open structure.

My trial account on blob storage is only one month. I'm wondering whether there is a way for me to apply for some storage resources to experiment those ideas. Also, is there any development like geemap<https://geemap.org/> for PC that we can publish the maps generated by PC? Do you have something like PC academic advocators who can help test, use, and advocate the platform?

Sorry for such a long email.

Thanks,
Xingong

**Reply by TomAugspurger** [reply]
> I'm wondering whether there is a way for me to apply for some storage resources to experiment those ideas.

Not at the moment, but possibly in the future.

> Also, is there any development like geemap<https://geemap.org/> for PC that we can publish the maps generated by PC?

I'm not familiar with geemap, so I'm not sure. But if you have COGs in blob storage, there are tools like titiler or leafmap that can visualize them.

> Do you have something like PC academic advocators who can help test, use, and advocate the platform?

Nothing formal, no.

**Comment by XingongLi** [comment]
Tom,

I was trying to compare the running time between 'local' file system, blob storage and parallelized blob storage using 3 test watershed from small to large . However, my PC account didn't allow me to upload all 3 watershed data to my local account as the small and medium watershed data have used up my disk quota. Is it possible to increase my disk quota so that I can finish the test and comparison?

So far, with just the small and medium watersheds, I found using blob storage is about 30 to 50% slower than using the local file system. But parallelized version (with 20 workers) only used about 35% of the time of local file system. I think I may further improve the parallelization to save more time.

Thanks,
Xingong

From: Li, Xingong
Sent: Friday, February 11, 2022 10:49 AM
To: microsoft/PlanetaryComputer ***@***.***>
Subject: RE: [microsoft/PlanetaryComputer] Access to local files by Dask Gateway workers (Discussion #31)

Tom,

I've got those things figured out. I changed my code and uploaded one small watershed to azure blob storage for testing. The parallelized flood mapping was running last night and used only half of the time of a linear mapping! I'll try a large watershed (I hope it won't reach my storage limit if there is one) and see the benefit of parallelization. I'm very new to Dask and would like to explore more on better the parallelization.

In addition to the flood mapping project, I also plan to do a global snow cover change analysis on PC using MODIS MOD10A1 & MYD10A1 data. I have done some before on Earth Engine (see TrendySnow<https://trendysnow.herokuapp.com/> web application built using EE as the backend engine). But there are limitations on using some trend analysis methods on EE. I think I may be able to do those analyses on PC with its open structure.

My trial account on blob storage is only one month. I'm wondering whether there is a way for me to apply for some storage resources to experiment those ideas. Also, is there any development like geemap<https://geemap.org/> for PC that we can publish the maps generated by PC? Do you have something like PC academic advocators who can help test, use, and advocate the platform?

Sorry for such a long email.

Thanks,
Xingong

**Comment by XingongLi** [comment]
I started using MODIS MOD10A1 and MYD10A1 daily snow cover data. It seems to me that MODIS data is still not in PC Catalog accessible through the STAC API. Any plan to move it into the Catalog? Also, has the data been moved to west Europe where the PC is running? Xingong

**Comment by XingongLi** [comment]
A few more questions:

1.       Is there a YAML file for the notebook env that I can recover. I installed a few packages and messed it up. Now I want to rollback to the original env.

2.       Is there a way that I can choose an env to run a notebook in the Hub?

3.       I'm trying to use PC through VSC following this doc<https://planetarycomputer.microsoft.com/docs/overview/ui-vscode/> but keeping getting the error "Failed to get remote kernel connections [FetchError: request to ***@***.***/api/kernels?1646287544781 failed, reason: certificate has expired". What certificate does it refer to? I have a token but am not sure what you mean "start a notebook server" in the instructions.

**Reply by TomAugspurger** [reply]
> I installed a few packages and messed it up. Now I want to rollback to the original env.

Restarting the notebook server should restore things, since changes outside of your home directory aren't persisted. Otherwise, the environments are defined in https://github.com/microsoft/planetary-computer-containers.

>  Is there a way that I can choose an env to run a notebook in the Hub?

JupyterHub environments / profiles, which determines the software environment and hardware, can only be chosen when logging in.

> I'm trying to use PC through VSC following this doc

Apparently, there's an additional setting that you need to configure. See https://github.com/microsoft/planetary-computer-hub/issues/45#issuecomment-1044608355. I haven't had a chance to debug that yet.
